src/OpenLoco/S5/SawyerStream.cpp

```diff
@@ -0,0 +1,488 @@
+}
+
+bool SawyerStreamReader::validateChecksum()
+{
+    auto valid = false;
+    auto backupPos = _stream.tellg();
+
+    _stream.seekg(0, std::ios::end);
+    auto fileLength = static_cast<uint32_t>(_stream.tellg());
+    if (fileLength >= 4)
+    {
+        // Read checksum
+        uint32_t checksum;
+        _stream.seekg(fileLength - 4);
+        _stream.read(reinterpret_cast<char*>(&checksum), sizeof(checksum));
+
+        // Calculate checksum
+        uint32_t actualChecksum = 0;
+        _stream.seekg(0);
+        uint8_t buffer[2048];
+        for (uint32_t i = 0; i < fileLength - 4; i += sizeof(buffer))
+        {
+            auto readLength = std::min(sizeof(buffer), fileLength - 4 - i);
+            _stream.read(reinterpret_cast<char*>(buffer), readLength);
+            for (size_t j = 0; j < readLength; j++)
+            {
+                actualChecksum += buffer[j];
+            }
+        }
+
+        valid = checksum == actualChecksum;
+    }
+
+    // Restore position
+    _stream.seekg(backupPos);
+
+    return valid;
+}
+
+void SawyerStreamReader::close()
+{
+    _stream.close();
+}
+
+stdx::span<uint8_t const> SawyerStreamReader::decode(SawyerEncoding encoding, stdx::span<uint8_t const> data)
+{
+    switch (encoding)
+    {
+        case SawyerEncoding::uncompressed:
+            return data;
+        case SawyerEncoding::runLengthSingle:
+            _decodeBuffer2.clear();
+            _decodeBuffer2.reserve(data.size());
+            decodeRunLengthSingle(_decodeBuffer2, data);
+            return _decodeBuffer2.getSpan();
+        case SawyerEncoding::runLengthMulti:
+            _decodeBuffer2.clear();
+            _decodeBuffer2.reserve(data.size());
+            decodeRunLengthSingle(_decodeBuffer2, data);
+
+            _decodeBuffer.clear();
+            _decodeBuffer.reserve(_decodeBuffer2.size());
+            decodeRunLengthMulti(_decodeBuffer, _decodeBuffer2.getSpan());
+            return _decodeBuffer.getSpan();
+        case SawyerEncoding::rotate:
+            _decodeBuffer2.clear();
+            _decodeBuffer2.reserve(data.size());
+            decodeRotate(_decodeBuffer2, data);
+            return _decodeBuffer2.getSpan();
+        default:
+            throw std::runtime_error(exceptionUnknownEncoding);
+    }
+}
+
+void SawyerStreamReader::decodeRunLengthSingle(FastBuffer& buffer, stdx::span<uint8_t const> data)
+{
+    for (size_t i = 0; i < data.size(); i++)
+    {
+        uint8_t rleCodeByte = data[i];
+        if (rleCodeByte & 128)
+        {
+            i++;
+            if (i >= data.size())
+            {
+                throw std::runtime_error(exceptionInvalidRLE);
+            }
+
+            auto copyLen = static_cast<size_t>(257 - rleCodeByte);
+            auto copyByte = data[i];
+            buffer.push_back(copyByte, copyLen);
+        }
+        else
+        {
+            if (i + 1 >= data.size() || i + 1 + rleCodeByte + 1 > data.size())
```

Comment:
if we did an i++ like L238 couldn't we remove all of the extra `+1`'s. Could even move the i++ outside of the if and into L235.